### ISSUE TITLE
Add information for uploading files

### ISSUE THREAD

**Issue by Ad-Alf** [jpickup/GarminTools#13]
Please add information on how and where to upload the files.

Following the developer guide on [FIT SDK Overview](https://developer.garmin.com/fit/overview/) I place the generated files in `Garmin\NewFiles`. 
After disconnecting my watch, I have the workouts on the watch and I am able to start them, but when I want to start them from the schedule, I have empty workouts.

**Comment by jpickup** [comment]
Yes, Garmin\NewFiles is correct. I've not experienced the problem you are seeing so I'm not quite sure how to reproduce it.

A few things to point out though:
Garmin limits the number of workouts that you can have in a schedule. I'm not sure what the limit is, but for me it amounted to about a month of workouts on a marathon plan. The software therefore only includes future planned workouts in the schedule such that if you re-generate and upload every month there are always planned workouts to follow (bizarrely this wasn't a limit on the old Forerunner 305 - it's something that Garmin added to newer watches)

I'd therefore suggest, as a test, to create a basic schedule with say 3 or 4 scheduled workouts over the next week or so and see how that looks. 

Alternatively, please send me an Excel file that exhibits the problem and I'll see if I can work out what's going wrong.

**Comment by jpickup** [comment]
I have just updated the instructions as per your suggestion.

**Comment by Ad-Alf** [comment]
I'm using a Forerunner 965. I created a spreadsheet with 2 workouts, just to test the app before creating an entire schedule.

Below you can find the excel I used. Perhaps good to mention that I created it in Google Spreadsheets.
attached pictures are how my agenda looks on my watch and how the workout looks when I select it in my workout calendar. If I want to start the workout, I have to select an activity and running isn't one of them.


![IMG_1748](https://github.com/user-attachments/assets/a9c35185-2140-4b96-ab26-e2b4b2e52303)
![IMG_1749](https://github.com/user-attachments/assets/1a85cb04-5e8b-4e1c-b44a-04ef4d9bde0f)
[Mijnentocht.xlsx](https://github.com/user-attachments/files/17893808/Mijnentocht.xlsx)

**Comment by Ad-Alf** [comment]
Deleted the workout .fit files and started all over again with the generation and it seems to work now.

**Comment by jpickup** [comment]
I'm glad it's working for you now, but I am worried that there is some issue that might affect others.

One (bad) thing that I have done with this software is hardcode Running as the activity type, as that's what I use the watch for. I almost certainly need to have this as either something in the spreadsheet( not very nice) or as a drop-down selection on the app. 

It's therefore possible that by doing this I've created something that doesn't work for other activity types.  Were you trying to use this for running or something else?

**Comment by Ad-Alf** [comment]
I used it for running.

After running it as administrator, it worked. So I believe that is the reason of the issues before.

Verzonden vanuit Outlook voor iOS<https://aka.ms/o0ukef>
________________________________
Van: John Pickup ***@***.***>
Verzonden: Sunday, November 24, 2024 6:07:57 PM
Aan: jpickup/GarminTools ***@***.***>
CC: Ad-Alf ***@***.***>; State change ***@***.***>
Onderwerp: Re: [jpickup/GarminTools] Add information for uploading files (Issue #13)


I'm glad it's working for you now, but I am worried that there is some issue that might affect others.

One (bad) thing that I have done with this software is hardcode Running as the activity type, as that's what I use the watch for. I almost certainly need to have this as either something in the spreadsheet( not very nice) or as a drop-down selection on the app.

It's therefore possible that by doing this I've created something that doesn't work for other activity types. Were you trying to use this for running or something else?

**Comment by florianleux** [comment]
On the topic of visualizing the workout, I uploaded my fit files and I indeed see the training for today.

But is there a way to see the coming workouts ? When I check the calendar in Garmin connect app I cannot see them, and if I go to the "training" section in the app I cannot see the trainings definitions.

**Comment by jpickup** [comment]
The workouts and schedule only go to the watch and, unfortunately, that doesn't sync back to Garmin Connect. The watch should however show you the future workouts.

One thing in the Garmin Tools app is an option to convert your workout plan to an iCal file, which can then be loaded into a calendar app. I used this to populate the calendar so I could easily see the future workouts on my phone.

Interaction with Garmin Connect requires a license from Garmin, which as an amateur developer is too expensive.

**Comment by Abasz** [comment]
Just FYI, there are several workflows that mock a full garmin connect login sequence. 

I.e. its not through the API key but rather fakes a headless browser login gets the token and makes calls to the API with that token. of course you need to rate limit the number of requests and things like that but it is available for almost every language (including dotnet: https://github.com/sealbro/dotnet.garmin.connect or this https://github.com/lswiderski/yet-another-garmin-connect-client).

Even I had a forked one for C# that worked at some point in time (https://github.com/Abasz/garmin-connect-client) :) There are newer ones that can do two factor auth as well with garmin.

Now I use a nodejs app for uploading workouts. So technically integration of direct upload to the calendar which is then synced to the watch is definately possible. Of course its a question whether it makes sense to add this feature, especially as you would need to create the workout json schema that the API requires.

**Comment by jpickup** [comment]
Thanks @Abasz I have been thinking that I'd like to port this project to some kind of web service and this is probably the way to do it. Do you have more info on the nodejs app? My Node and Javascript skills aren't great but this could be a good opportunity to learn!

**Comment by surbeym** [comment]
Would happy to help with this! Big in the spring boot world, and some react
and vue skills

**Comment by Abasz** [comment]
> Thanks @Abasz I have been thinking that I'd like to port this project to some kind of web service and this is probably the way to do it.

I have a nodejs app which uses nest.js framework for backend and angular front to run several Garmin related things: backup activities, convert some fit files, and manage workouts (no calendar). This is private project, but I can clean it up and share for reference. Note that this has its own parser for the workouts (in regex, but it is based very much on your parser's language). It does not handle 2FA login unfortunately (the time I worked this out it was not a thing and then I did not spend time to update) but there a re library that would.

This downloads activities and extracts recovery heart rate and plots them so  I can track progress of my fitness
![Image](https://github.com/user-attachments/assets/c001a01c-1b42-4c36-91a6-8401633bef7b)

This is just because for my kayaking I record cycling (long story short garmin only allows certain sensors, like cadence in certain activities and on my paddle I record strokes via cycling cadence BLE) and convert it to rowing.
![Image](https://github.com/user-attachments/assets/76b24f53-c6d5-48a7-8bc1-50a2525b91f8)

This is the workout manager. There is the "language" and than you can sync all workouts or individual ones. If workout exists it updates if not it creates in Garmin Connect. No Calender, for me that was not necessary
![Image](https://github.com/user-attachments/assets/c946fb46-4917-4c2e-b908-b7aab04b4b42)

![Image](https://github.com/user-attachments/assets/8d006e03-1009-4bdb-ac7c-b39bfd3c88b5)

The whole thing work in a way that by launching an executable it starts the backend nodejs server (that runs in the console) and you access from the browser via entering the ip of the server... Not ideal for a standalone app... I always wanted to move it .NET Maui (so it work in phones etc.) or electron so it has this native app look, but I never did, mostly because the GUI code would require a lot of work...

Any way, having this garmin connect connection service is not really for a web service run on a central server accessible by others. Reason is that Garmin API, without the API key does not tolerate a lot of logins with different credentials and too many requests at once (protected by cloudflare). Not to mention that you need to enter/send the password in plain text, such central web service used by others would be able to see that. So for this reason it is better if the app runs on the user's machine and everyone uses it on their own.

In this sense, a native app, in my view is a better approach, that logs the user in. Actually the login can be faked by any language (I have seen python, .net, nodejs, go etc.) code that does the same thing. One thing that is required is some auth token for cloudflare that someone has published (https://github.com/petergardfjall/garminexport/issues/102)

I can clean up my code from security concerns like hard coded things and share with you if you want. The typings for the upload object and a like can be useful.

But to be fair, I think nodejs is not the right tool for this job. But it is true that I have not found a Java implementation for the Garmin Client (though I have not searched too hard :)). but you can try to port it, the logic is not too complicated: the most important parts are cookies that one need to acquire and keep in the jar for the subsequent request.
